### DISCUSSION TITLE
Availability of Drivers Windows 10 x64 & ARM64 too

### DISCUSSION THREAD

**Discussion by Corsair-63** [peterantypas/maiana#27] [Q&A]
It would be nice that would be available drivers for the connection of a computer running windows 10 either x64 either ARM64 (Surface Pro X ver 2019 & 2020), these last are growing in the tablet area and really very useful on board.

**Comment by peterantypas** [comment]
What interfaces do these have that would be suitable? How do they connect to other boat systems today?

**Reply by marcelrv** [reply]
Here : wireless (esp8266) and good old nmea over serial port.
If you don't want to build WiFi in all module s it would great if the control box would have a to place to ( pinheader) to add one.

**Reply by peterantypas** [reply]
What signals out of the control box would you need to support it?

**Reply by Corsair-63** [reply]
As far as I see, all the signal, excluding network (wired-wireless), including BT are done basis in a COM port because the receiver PC, receives the signal via such interface so the input signal at least has to be recognises as a COM port, unless you send the signal in NMEA+/- wiring to a multiplexor connected to another devices and connected to the computer (but also this connection must be recognised by PC as a COM port).

and even wired-wireless signals or the PC navigation software if it doesn't accept network inputs, they have to be re-routes to a converted network-virtual COM port, this has been a real nightmare for me with my tablet (and still not able to do it), for computers W10x64 based there is not so much problem, there are a lot of converter software.

I included in my question/idea the version of COM port for windows x64, because version x86 is being obsolete, and also W10ARM64, because tablets based in ARM are growing and expected to grow even  more ARM based PC (or is what computing gurus are predicting)
and having drivers for COM ports in ARM64 is becoming a nightmare, because such version only accept drivers specifically built for it, it doesn't accept inherited ones from versions x86 nor x64.

**Reply by marcelrv** [reply]
I'm using openplotter & signalK which happily accepts serial over ethernet without converting it first to a comport. 
A header for esp module would allow to add wifi as well as Bluetooth connectivity for less than than $5. Likewise navionics, nvcharts on tablet only has wireless options.

**Reply by peterantypas** [reply]
Does anyone know if Windows ARM supports a standard USB CDC (comport) interface as implemented in some STM32 series? This could be a solution, but it would force the use of an MCU in the control box.

**Reply by Corsair-63** [reply]
All adaptors I've tried for connecting one COM port to Surface Pro X (2019-W10ARM64) none of them had been able to set-up in the tablet because lack of drivers specifically built for ARM64, connected to another computers with also windows 10 x64 (Intel either AMD) drivers directly downloaded from Microsoft and smoothly running in seconds.

I've tried with:
- Eltima Software, no available 
- Interfaz NMEA 0183 A USB Adaptador Cable Digital Yacht (FTDI interface), no way
- Digital Yacht uAIS Receiver (FTDI interface), no way
- and another options but never found any.

in some cases the only option found was via a Bluetooth COM port when possible to re-route de data stream

**Reply by astuder** [reply]
> What signals out of the control box would you need to support it?

The pinouts of HC-06/05 BT, HC-08 BLE and DT-06 WiFi modules are compatible, though sometimes mirrored. Only the center 4 pins (VCC, GND, TX, RX) are needed.

When adding a header/pads to the PCB, keep in mind that these modules typically come with right-angle headers soldered.

As most ARM64 devices are tablets, I'd go with a wireless connection. I don't know of any navigation app that supports AIS over serial but not network. And in my experience with Android tablets, the even bigger headache is to power/charge the tablet while connected to a USB device - not all tablets support that.

**Reply by peterantypas** [reply]
We're on the same page here. Tablets are meant to be wireless. That said, I'm going to take the boat for a float today (it's nice out) and I'll re-think the interface. Maybe it makes sense to break it into 2 parts: Control/power + adapter and have a variety of plugin adapters for maximum compatibility.

**Reply by astuder** [reply]
> Does anyone know if Windows ARM supports a standard USB CDC (comport) interface as implemented in some STM32 series? This could be a solution, but it would force the use of an MCU in the control box.

In my experience, all STM32 VCOM devices required a driver from ST. But there's probably a plain USB CDC example somewhere in their SDK. Standard USB CDC on Windows 10 should work without driver independent of the platform, but you will require your own HID/PID.

And again as an extra headache for Android tablets: Applications need to explicitly support your HID/PID for serial connectivity over USB CDC.

**Reply by Corsair-63** [reply]
> What signals out of the control box would you need to support it?<

TX,RX,GND,VCC
all data from GNSS uses to be on 4800bps n81p
All data from AIS that uses to be on 38400 n81p
on that way could be interfaced to any PC via COM port either physically either BT (FTB only one working wit W10ARM64)

> In my experience, all STM32 VCOM devices required a driver from ST. But there's probably a plain USB CDC example somewhere in their SDK. Standard USB CDC on Windows 10 should work without driver independent of the platform, but you will require your own HID/PID. <

that's the main problem of W10ARM64 it only accepts drivers compiled for it, It doesn't accept the ones done for x86 nor x64 unless they had been compiled as UWT, and of course there aren't so much.

**Reply by peterantypas** [reply]
The MAIANA stream is 38400 n81 and that includes GPS as well.

**Reply by Corsair-63** [reply]
> The MAIANA stream is 38400 n81 and that includes GPS as well.

of course that is accepted by many software and it's how I have configured in some of them.

FTB, only one that need separate AIS entrance that I've tested is TIKI NAVIONICS GOLD 3.2, nice one but sometimes tricky to set-up but very good support of all devices connected, with above restriction.

**Reply by peterantypas** [reply]
> > What signals out of the control box would you need to support it?
> 
> The pinouts of HC-06/05 BT, HC-08 BLE and DT-06 WiFi modules are compatible, though sometimes mirrored. Only the center 4 pins (VCC, GND, TX, RX) are needed.
> 
> When adding a header/pads to the PCB, keep in mind that these modules typically come with right-angle headers soldered.
> 
Easy, a horizontal 0.1" female socket will do this. So looking at these devices you listed, I can support all except the HC-08 because it has an incompatible pinout (RX & TX are swapped). For the others, I can leave room for the outer pins and supply 5V. How are these used to connect to apps on tablets?

**Reply by astuder** [reply]
Sorry, looks like HM-10 is the compatible BLE module. See "HM-10 breakout board" on this page:
http://www.martyncurrey.com/hm-10-bluetooth-4ble-modules/

Anyways, WiFi is what most people would use these days. These modules just output what they receive on RX, i.e. for AIS it's the NMEA data as you'd get on serial. The WiFi modules have a web UI to configure baud rate and network mode (TCP server, UDP broadcast etc.). 

A challenge with the mirrored pin-out is that the module will sit below the plane of the PCB, i.e. the header must face towards an edge, or a cutout in the PCB. The image below shows a HC-06 and a DT-06 module. Over time I've received mirrored and non-mirrored versions of the DT-06 from the same vendor.
![dAISy2PlusSerial](https://user-images.githubusercontent.com/1166488/102940723-d2907400-4465-11eb-9090-56a6fb3004e8.jpg)

The easiest is to have an external "expansion port" like I did with the dAISy 2+, but it would be more elegant if the module were contained within the (non-metallic) enclosure.

**Reply by astuder** [reply]
PS: The configuration website on the DT-06 is pretty ~~simple~~ technical, and I doubt very secure. An alternative approach would be a socket for a popular ESP module and then let the community build an app more appropriate for the task. Similar to what NMEA2WIFI did:
https://www.vela-navega.com/index.php/nmea2wifi

**Reply by peterantypas** [reply]
Thanks. It looks like the female header needs to be flush with the enclosure for them to be snug. I'm going to order some of these and go through a fitting exercise. I will probably move all the outputs to the south side of the enclosure to de-clutter the north side. Adding the pin header is trivial, the issue in my case is I can't have parts protruding from the PCB on both sides as it will make assembly impossible.

I don't think I'll be able to fit them inside without some serious layout challenges either. Plus, I would have large metal switches at close proximity to the antenna.

**Reply by peterantypas** [reply]
Looking at the two samples of these modules I got, they obviously don't have the same module in them (only one displays an FCC ID). What is a "current" version of the core module (the one soldered onto the breakout as pictured above) that I may want to integrate into the breakout box directly? I'll have to redesign that PCB for RF now, but it may in fact be the right thing to do.

**Reply by peterantypas** [reply]
Here is what I got from the FCC for ID 2AL3B-ESP-M:

[ESP32-ExternalPhotos.pdf](https://github.com/peterantypas/maiana/files/5760704/ESP32-ExternalPhotos.pdf)
[ESP32-InternalPhotos.pdf](https://github.com/peterantypas/maiana/files/5760705/ESP32-InternalPhotos.pdf)

Is this module "current"? Can it be easily obtained and configured with stock firmware? I'm not familiar with this platform at all, as you may have guessed ;)

The breakout doesn't seem to add much value outside of an LDO and two buttons (presumably for reset/boot mode?)

**Reply by marcelrv** [reply]
Yes, these are the 'bare modules'. These would be good if you really redesign the board to make place for it and a pcb antenna.

However mostly folks would buy the ones with pins
like this https://nl.aliexpress.com/item/32816065152.html

Dunno if you ever watched this guy, but he has a great overview of current modules.
https://www.youtube.com/watch?v=mnoZYlyebBc

**Reply by astuder** [reply]
I usually buy them from BangGood. The version with the FCC number came from there. 
https://www.banggood.com/DT-06-Wireless-WiFi-Serial-Port-Transparent-Transmission-Module-TTL-To-WiFi-With-bluetooth-HC-06-Interface-ESP-M2-p-1141047.html

Last summer I bought a bag of them from what I think is DOIT's official store on AliExpress. That's the version without FCC number.
https://www.aliexpress.com/item/32813585141.html

Note that the images on the two websites show mirrored modules. I'm pretty certain that I received different versions from BangGood over the years. 

The benefit of using these over a generic ESP module is, that the end-user doesn't have to worry about programming them. Though I heard from people that bought them off Ebay with mixed results, i.e. non-working modules. I suspect those weren't programmed with the correct WiFi UART bridge firmware.

**Comment by Corsair-63** [comment]
Most of the times as per your message:

> The pinouts of HC-06/05 BT, HC-08 BLE and DT-06 WiFi modules are compatible, though sometimes mirrored. Only the center 4 pins (VCC, GND, TX, RX) are needed.

and now depends of the interface you want to do, if you send to a multiplexer, just with a 4 core cables, if you want to connect straight away to a computer with a NMEA>Com port (USB) converted like Digital Yacht one with the cables in the unit and the USB to the PC.

and of course the transceiver unit had to send the data in a NMEA known format as RS232 either RS422, most easy is RS232 and sometimes compatible.

**Comment by peterantypas** [comment]
I got samples of these modules, and here is what I think makes the most sense:

![Screenshot from 2020-12-28 19-58-58](https://user-images.githubusercontent.com/1565933/103257727-3e337f00-4947-11eb-9936-090df3582d8c.png)
![Screenshot from 2020-12-28 19-58-40](https://user-images.githubusercontent.com/1565933/103257730-3f64ac00-4947-11eb-8eb7-938226adbb4c.png)

This layout moves all outputs (USB + RF) to the bottom. The meandering trace antenna would be partially exposed.

**Reply by marcelrv** [reply]
I like the move to the bottom.
 Still struggling why we would keep the antenna trace exposed Vs having the whole module inside. I would expect the plastic enclosure would not be a problem.
Having the module completely inside looks better optically and maybe bit better protected against humidity

**Reply by peterantypas** [reply]
Three reasons:

- There are 3 large metal objects inside which will be milimeters from the antenna no matter how it's oriented
- There must be no PCB ground plane anywhere near a meandering trace antenna
- These modules are often supplied inverted in a random / unpredictable way, thus requiring "upside down" mounting

**Reply by peterantypas** [reply]
An alternative might be to simply have a female expansion header exposed at the bottom and let the end user wire / plug whatever system they want. This would make it completely external but also offer the most flexibility.

Fully encasing a wireless module requires design from scratch. You can't just plop an antenna into an existing design and just hope it works. It almost never does.

**Reply by PVHL** [reply]
Just want to say hi. I've been noticing the activity since I checked this out a while back and it didn't look active. Where do I look for current status, roadmap? Not sure what you're currently aiming at, open design or finished product. Am very much hoping to be able to have AIS tx/rx for my Canadian boat from an open source. My background is in electronics (mostly fixing and from a while back) and programming, among other things (like music and ministry). Cheers.

**Reply by marcelrv** [reply]
> An alternative might be to simply have a female expansion header exposed at the bottom and let the end user wire / plug whatever system they want. This would make it completely external but also offer the most flexibility.

Yes, that would indeed work

**Reply by peterantypas** [reply]
@PVHL My goal is to enable individuals with good technical skills to build their own AIS transponder for their boats. That said, what constitutes "building your own" transmitter is open to a lot of interpretation, and that's exactly where I plan to operate. You could build the system on your own, or I can sell you a kit that you still have to finish yourself. I don't know exactly how long it will take me to get there, but I'm hoping to be up and running before the next boating season starts. Main challenge right now is supply chain and the fact that I just bought a condo so I need to pack and move.
